designs/0013-profiling.md

```diff
@@ -0,0 +1,231 @@
+
+Examples of cases where evaluating is very useful: evaluating paralleliziation approaches (reduce_sum, map_rect, GPUs), identifying the effect of using a different solver or a different expression.
+
+# Profiling on the Stan model level
+[stan-model-profiling]: #stan-model-profiling
+
+A user defines profiling sections inside a model by writing `profile("profile-section-name");`. The argument to `profile()` is always a string literal. A profile will measure execution time and other AD information for all expressions for the Stan lines for which the profile is in scope.
+
+An model with examples of its usage is shown below:
+
+```stan
+data {
+  int<lower=1> N;
+  real x[N];
+  vector[N] y;
+}
+transformed data {
+  profile("tdata");
+  vector[N] mu = rep_vector(0, N);
+}
+parameters {
+  real<lower=0> rho;
+  real<lower=0> alpha;
+  real<lower=0> sigma;
+}
+transformed parameters {
+  matrix[N, N] L_K;
+  matrix[N, N] K;
+  {
+    profile("gp");
+    K = gp_exp_quad_cov(x, alpha, rho);
+  }
+  real sq_sigma = square(sigma);
+
+  for (n in 1:N) {
+    profile("add_diagonal");
+    K[n, n] = K[n, n] + sq_sigma;
+  }    
+
+  profile("cholesky");
+  L_K = cholesky_decompose(K);
+}
+model { 
+  {
+    profile("priors");
+    rho ~ inv_gamma(5, 5);
+    alpha ~ normal(0, 1);
+    sigma ~ normal(0, 1);
+  }
+  
+  {
+    profile("likelihood");
+    y ~ multi_normal_cholesky(mu, L_K);
+  }
+}
+```
+
+The model uses six profiles: 
+- `tdata` profiles the only statement in the transformed data block
+- `gp` profiles the `gp_exp_quad_cov` function
+- `add_diagonal` profiles the for loop. Profiles in a loop will accumulate as will profiles with the same name on different lines.
+- `cholesky` profiles the Cholesky decomposition as it goes out of scope at the end of the transformed parameters block.
+- `priors` profiles the three prior definitons in the model block
+- `likelihood` profiles the `multi_normal_cholesky_lpdf` function
+
+There are no other changes to Stan models or language. 
+
+Note that we can not profile the following as only parameter declarations are allowed in the `parameters` block.
+```stan
+parameters {
+  real<lower=0> rho;
+  real<lower=0> alpha;
+  real<lower=0> sigma;
+}
+```
+This means that the `lower`, `upper`, `multiply` and `offset` constraints/transforms can not be profiled. Given that these are less likely to be a bottleneck we decided to not change the Stan language rules to allow `profile()` in the parameters block for only this use-case.
+
+# Stanc3 - changes to the generated C++
+
+The C++ code generated from Stan models will be changed accordingly:
+
+- all generated C++ models representing Stan models will gain a private member `profiles__`. This member will represent the map of all profile information.
```

Comment:
Alright so the profile is a member variable of a model. When is the profile zero'd out?

Like if I do two sampling-s in Rstan (when it gets to 2.26), would the profile get reset? And what about if someone is using the `grad` calls in Rstan?

Reply:
Hm, I need to check what rstan does. Does it call the model constructor for each sampling call? I would say it does as the constuctor handles input data. If it does not it would have to be reset.